File: 20. Valid Parentheses.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for(int i=0;i<s.length();i++){
            Character current = s.charAt(i);

            if( current.equals('(') || current.equals('{') || current.equals('[')){
                stack.push(current);
            }
            else{
                Character lastOpen;
                if(stack.empty()){
                    return false;
                }
                else{
                    lastOpen = stack.peek();
                }

                if(current.equals(')') && !lastOpen.equals('(')){
                    return false;
                }
                else if (current.equals(']') && !lastOpen.equals('[')){
                    return false;
                }

                else if (current.equals('}') && !lastOpen.equals('{')){
                    return false;
                }

                stack.pop();
            }
        }

        return stack.isEmpty();

    }
}

File: 20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = collections.deque()

        for ch in s:
            if ch == "(": stack.append(")")
            elif ch == "[": stack.append("]")
            elif ch == "{": stack.append("}")
            else: 
                if not stack or stack.pop() != ch: return False 
    
        return not stack